dbms/src/Flash/Mpp/ExchangeReceiver.cpp

```diff
@@ -945,13 +911,24 @@ ExchangeReceiverResult ExchangeReceiverBase<RPCContext>::handleUnnormalChannel(
 
 template <typename RPCContext>
 ExchangeReceiverResult ExchangeReceiverBase<RPCContext>::toDecodeResult(
+    size_t stream_id,
     std::queue<Block> & block_queue,
     const Block & header,
-    const RecvMsgPtr & recv_msg,
+    const ReceivedMessagePtr & recv_msg,
     std::unique_ptr<CHBlockChunkDecodeAndSquash> & decoder_ptr)
 {
     assert(recv_msg != nullptr);
-    if (recv_msg->resp_ptr != nullptr) /// the data of the last packet is serialized from tipb::SelectResponse including execution summaries.
+    bool handle_resp = false;
+    if (recv_msg->remaining_consumers != nullptr)
+    {
+        /// fine grained shuffle
+        handle_resp = stream_id == 0 && recv_msg->resp_ptr != nullptr;
+    }
+    else
+    {
+        handle_resp = recv_msg->resp_ptr != nullptr;
+    }
```

Comment:
Looks like it can be wrapped into a function.
```
ReceivedMessage::respPtr(size_t stream_id) { return stream_id  == 0 ? resp_ptr : nullptr;  }
```
Because the stream_id of non-fine-grained shuffle must be 0, so there is no need to judge remaining_consumers here?

Reply:
Ok, I've made `ReceivedMessage` as a class, and added some member functions.